### PR TITLE
[RF] add LikelihoodJob support and some RooFit::TestStatistics fixes

### PR DESCRIPTION
**Pull request by egpbos**
# This Pull request:

## Changes or fixes:
The main feature is that this PR makes it possible (in the final commit 5227996a9941918404d0a36037daa2c91e29833a) to run minimizations with both LikelihoodGradientJob and LikelihoodJob activated at the same time. This adds likelihood parallelization at the event-level to the non-gradient phases of Minuit migrad minimization: 1. the setup phase where the initial gradient is calculated and 2. the line search phases.

To make the above possible, a few bugs had to be fixed:
- `RooAbsL` classes have `evaluatePartition` which allows you to specify the range of events and components over which to evaluate the likelihood. This had some bugs: the extended term and possible subsidiary components were added for each partition, leading to N duplicates (for N partitions). Commits 3db2e0dd36478813cfb534451f459348e8d90da9 and c2a8bc9df029f65ee575e140825af6c509860538.
    - As a result of this fix, which reorders some terms in the sum, the LikelihoodGradientJob test was failing for a specific case due to increased floating point differences. We would like to add back in an option for retaining bit-wise exact same results in the future, but since the minimization still converges to the same value within the desired precision, for now we leave it like this, since the user can still pick the old ways if they prefer their exact old results. Commit b7f7fa7b705e2dfb767187dce3f6a32275f30d9e.
- `LikelihoodJob` had a bug where it didn't reset its `result` sum variable to zero before doing a new sum. Commits 042e1577d8c303d532016537e7467d6e2fbfcd9c and 44197a2a64f77b47587e5ee200d360c4514b8098.

Finally, I made a mini optimization in `LikelihoodJob` by using `publish_from_master_to_workers` more efficiently, which reduces communication overhead slightly (and makes the code a bit cleaner). Also in commit 567b7988d56b6c1db6102d857b523fc2da607dbc.

Note that although this is a bit of a big PR, the commits are factored out into the above list of changes, so I hope this makes it easy to review.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

Documentation not yet updated, I will check if this is necessary (I didn't think so, because it doesn't actually add much to the user interface, all the classes were already there, they just didn't work before like this).

### CODE REVIEW

**Comment by guitargeek** [comment]
Hi, thanks for the PR! I suppose the test failures are not expected?

**Comment by egpbos** [comment]
Indeed, they are not expected... I'll need to look into it.

**Comment by egpbos** [comment]
Ok, two things:

1. Commit c680b3f adds some modifications that somehow got lost in my rebasing; this now makes everything pass on my laptop again.
2. ... however, this was still based on master from 40620e494ec. After rebasing on 70f96ca797b, tests now fail on my laptop as well.

So, it seems that something else in RooFit changed in the meantime.

**Comment by egpbos** [comment]
I don't want to speak to soon, but I think I fixed the issue @guitargeek, so feel free to review now.

Note that the PR still has two fixup commits right now that I will squash with the right ones in the history after CI passes on Mac and Windows too (don't want to trigger the rebuild and get put in the queue again ;)).

File: roofit/roofitcore/inc/RooFit/TestStatistics/RooAbsL.h [hunk 1]
```diff
@@ -76,6 +74,14 @@ class RooAbsL {
          }
       }
 
+      bool operator==(const Section& rhs) {
+         if (begin_fraction == rhs.begin_fraction && end_fraction == rhs.end_fraction) {
+            return true;
+         } else {
+            return false;
+         }
```
Comment (guitargeek) [inline roofit/roofitcore/inc/RooFit/TestStatistics/RooAbsL.h:82]:
```suggestion
         return begin_fraction == rhs.begin_fraction && end_fraction == rhs.end_fraction;
```
Maybe do this to save a few lines?

File: roofit/roofitcore/src/TestStatistics/MinuitFcnGrad.cxx [hunk 2]
```diff
@@ -74,8 +85,14 @@ double MinuitFcnGrad::DoEval(const double *x) const
...
    // Calculate the function for these parameters
    //   RooAbsReal::setHideOffset(false);
-   likelihood->evaluate();
-   double fvalue = likelihood->getResult().Sum();
+   std::shared_ptr<LikelihoodWrapper> likelihood_here;
+   if (gradient->isCalculating()) {
+      likelihood_here = likelihood_in_gradient;
+   } else {
+      likelihood_here = likelihood;
+   }
```
Comment (guitargeek) [inline roofit/roofitcore/src/TestStatistics/MinuitFcnGrad.cxx:93]:
```suggestion
   likelihood_here = gradient->isCalculating() ? likelihood_in_gradient : likelihood;
```

File: roofit/roofitcore/src/TestStatistics/RooSubsidiaryL.cxx [hunk 3]
```diff
@@ -55,20 +55,20 @@ RooSubsidiaryL::RooSubsidiaryL(const std::string &parent_pdf_name, const RooArgS
...
-                                            << std::endl;
-   }
-
    ROOT::Math::KahanSum<double> sum;
 
-   for (const auto comp : subsidiary_pdfs_) {
-      sum += -((RooAbsPdf *)comp)->getLogVal(&parameter_set_);
+   if (events.end_fraction == 1) {
+      for (const auto comp : subsidiary_pdfs_) {
+         sum += -((RooAbsPdf *)comp)->getLogVal(&parameter_set_);
```
Comment (guitargeek) [inline roofit/roofitcore/src/TestStatistics/RooSubsidiaryL.cxx:70]:
```suggestion
         sum += -static_cast<RooAbsPdf*>(comp)->getLogVal(&parameter_set_);
```
Can you change to C++-style cast here when already touching this line?

**Review by guitargeek — CHANGES_REQUESTED** [review]
Hi @egpbos, great that it seems to works now!

I have only really minor comments. Besides the ones I made inline, could you please reorder your commits a bit? There are commits that introduce failing tests that are fixed in the next commit, but it would be better to have it the other way around such that ROOT builds and tests are green for all commits in the history.

In any case, thanks a lot for all these fixes and implementing also the unit tests!

**Comment by egpbos** [comment]
I'll squash the failing test commits, no problem (I thought it could be nice to have them separate so that the history also shows that the tests indeed fail without the fix in the next commit, but I understand that it may be too much).

**Comment by egpbos** [comment]
@guitargeek Applied all requested review suggestions and rebased into 4 more compact logical commits.

**Comment by egpbos** [comment]
Did one more rebase, to put the last commit first, because it actually fixes the earlier commit 22864d5a4430ef28270b2f309cd742ff1764a56e and is necessary for the tests in the following commits to pass.

**Comment by egpbos** [comment]
The failing tests are not due to this PR. I think this should be ready to merge @guitargeek :)

**Review by guitargeek — APPROVED** [review]
LGTM, thank you very much for these fixes and improvements!

Please just go ahead and merge :)

**Pull request merged**